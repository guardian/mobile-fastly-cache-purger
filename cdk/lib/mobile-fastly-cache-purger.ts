import type {GuStackProps} from '@guardian/cdk/lib/constructs/core';
import {GuStack} from '@guardian/cdk/lib/constructs/core';
import {GuLambdaFunction} from '@guardian/cdk/lib/constructs/lambda';
import {GuardianAwsAccounts} from '@guardian/private-infrastructure-config';
import type {App} from 'aws-cdk-lib';
import {CfnParameter, Duration, Fn} from 'aws-cdk-lib';
// eslint-disable-next-line import/no-namespace -- sdf
import * as iam from 'aws-cdk-lib/aws-iam';
import {Runtime} from 'aws-cdk-lib/aws-lambda';
// eslint-disable-next-line import/no-namespace -- sdf
import * as lambdaEventSources from 'aws-cdk-lib/aws-lambda-event-sources';
import {Topic} from 'aws-cdk-lib/aws-sns';
import {SqsSubscription} from "aws-cdk-lib/aws-sns-subscriptions";
// eslint-disable-next-line import/no-namespace -- sdf
import * as sqs from 'aws-cdk-lib/aws-sqs';
import {GuLambdaDockerFunction} from "./gu-lamba-docker-function";


export class MobileFastlyCachePurger extends GuStack {

	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const faciaID = this.stage == "CODE" ? "StorageConsumerRole-1JWVQ2NTELFT7" : "StorageConsumerRole-1R9GQEVJIM323";
		const buildId = new CfnParameter(this, 'BuildId', {
			type: 'String',
			default: 'dev',
			description: 'Tag to be used for the image URL, e.g. riff raff build id',
		}).value.toString();

		const executionRole: iam.Role = new iam.Role(this, 'ExecutionRole', {
			assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
			path: "/",
			inlinePolicies: {
				logs: new iam.PolicyDocument({
					statements: [
						new iam.PolicyStatement({
							actions: [ 'logs:CreateLogGroup' ],
							resources: [ `arn:aws:logs:eu-west-1:${this.account}:*` ]
						}),
						new iam.PolicyStatement({
							actions: [ 'logs:CreateLogStream', 'logs:PutLogEvents' ],
							resources: [ `arn:aws:logs:eu-west-1:${this.account}:log-group:/aws/lambda/*:*` ]
						})
					] }),
				Conf: new iam.PolicyDocument({
					statements: [
						new iam.PolicyStatement({
							actions: [ 'ssm:GetParametersByPath' ],
							resources: [ `arn:aws:ssm:${this.region}:${this.account}:parameter/cache-purger/${this.stage}` ]
						})
					] }),
				Assume: new iam.PolicyDocument({
					statements: [
						new iam.PolicyStatement({
							actions: ['sts:AssumeRole'],
							resources: [`arn:aws:iam::${GuardianAwsAccounts.CMSFronts}:role/facia-${this.stage}-${faciaID}`]
						})
					]
				})
			}
		})

		const handler: GuLambdaFunction = new GuLambdaFunction(this, 'mobile-fastly-cache-purger', {
			handler: 'PurgerLambda::handleRequest',
			functionName: `mobile-fastly-cache-purger-cdk-${this.stage}`,
			timeout: Duration.seconds(60),
			environment: {
				App: 'mobile-fastly-cache-purger',
				Stack: this.stack,
				Stage: this.stage,
			},
			runtime: Runtime.JAVA_11,
			app: 'mobile-fastly-cache-purger',
			fileName: `mobile-fastly-cache-purger.jar`,
			role: executionRole,
		});

		const imageRepositoryArn = Fn.importValue('mobile-fastly-cache-purger-repository-arn')
		const imageRepositoryName = Fn.importValue('mobile-fastly-cache-purger-repository-name')
		const lambdaFunction: GuLambdaDockerFunction = new GuLambdaDockerFunction(this, 'mobile-fastly-cache-purger-v2', {
			functionName: `mobile-fastly-cache-purger-cdk-${this.stage}-v2`,
			timeout: Duration.seconds(60),
			environment: {
				App: 'mobile-fastly-cache-purger-v2',
				Stack: this.stack,
				Stage: this.stage,
			},
			app: 'mobile-fastly-cache-purger-v2',
			repositoryArn: `${imageRepositoryArn}`,
			repositoryName: `${imageRepositoryName}`,
			imageTag: `${buildId}`,
			role: executionRole,
			memorySize: 1024
		});

		const dlq: sqs.Queue = new sqs.Queue(this, "frontsPurgeDlq")

		const queue: sqs.Queue = new sqs.Queue(this, "frontsPurgeSqs", {
			visibilityTimeout: Duration.seconds(70), 	//default for a queue is 30s, and the lambda is 60s
			deadLetterQueue: {
				maxReceiveCount: 3,
				queue: dlq,
			}
		});

		const frontsUpdateTopicName=
			this.stage == "CODE" ? "FrontsUpdateSNSTopic-RepwK3g95V3w" : "FrontsUpdateSNSTopic-kWN6oX2kvOmI";

		const frontsUpdateTopic = Topic.fromTopicArn(
			this,
			"FrontsUpdateSNSTopic",
			`arn:aws:sns:${this.region}:${GuardianAwsAccounts.CMSFronts}:facia-${this.stage}-${frontsUpdateTopicName}`
		)
		frontsUpdateTopic.addSubscription(new SqsSubscription(queue));

		const eventSource: lambdaEventSources.SqsEventSource = new lambdaEventSources.SqsEventSource(queue);

		handler.addEventSource(eventSource);
		lambdaFunction.addEventSource(eventSource);
	}
}
