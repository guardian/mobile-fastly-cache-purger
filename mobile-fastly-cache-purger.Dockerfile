FROM public.ecr.aws/lambda/java:11
COPY mobile-fastly-cache-purger.jar ${LAMBDA_TASK_ROOT}/lib/
CMD ["PurgerLambda::handleRequest"]
