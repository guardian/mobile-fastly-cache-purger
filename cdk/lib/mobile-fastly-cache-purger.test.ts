import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { MobileFastlyCachePurger } from "./mobile-fastly-cache-purger";

describe("The MobileFastlyCachePurger stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new MobileFastlyCachePurger(app, "MobileFastlyCachePurger", { stack: "mobile-fastly-cache-purger", stage: "TEST" });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
