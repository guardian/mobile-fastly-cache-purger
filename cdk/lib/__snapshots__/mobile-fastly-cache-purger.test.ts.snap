// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The MobileFastlyCachePurger stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ExecutionRole605A040B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "logs:CreateLogGroup",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:logs:eu-west-1:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":*",
                      ],
                    ],
                  },
                },
                {
                  "Action": [
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:logs:eu-west-1:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*:*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "logs",
          },
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "ssm:GetParametersByPath",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:ssm:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":parameter/cache-purger/TEST",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Conf",
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-fastly-cache-purger",
          },
          {
            "Key": "Stack",
            "Value": "mobile-fastly-cache-purger",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ExecutionRoleDefaultPolicyA5B92313": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/mobile-fastly-cache-purger/TEST/mobile-fastly-cache-purger/mobile-fastly-cache-purger.jar",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/mobile-fastly-cache-purger/mobile-fastly-cache-purger",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/mobile-fastly-cache-purger/mobile-fastly-cache-purger/*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "frontsPurgeSqsC6BB4572",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ExecutionRoleDefaultPolicyA5B92313",
        "Roles": [
          {
            "Ref": "ExecutionRole605A040B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "frontsPurgeDlq224A89E8": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-fastly-cache-purger",
          },
          {
            "Key": "Stack",
            "Value": "mobile-fastly-cache-purger",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "frontsPurgeSqsC6BB4572": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "frontsPurgeDlq224A89E8",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-fastly-cache-purger",
          },
          {
            "Key": "Stack",
            "Value": "mobile-fastly-cache-purger",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VisibilityTimeout": 70,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "frontsPurgeSqsMobileFastlyCachePurgerFrontsUpdateSNSTopic7A288BA5EDF5FFD8": {
      "DependsOn": [
        "frontsPurgeSqsPolicy6F57468C",
      ],
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "frontsPurgeSqsC6BB4572",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "Region": {
          "Ref": "AWS::Region",
        },
        "TopicArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":163592447864:facia-TEST-FrontsUpdateSNSTopic-RepwK3g95V3w",
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "frontsPurgeSqsPolicy6F57468C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:sns:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":163592447864:facia-TEST-FrontsUpdateSNSTopic-RepwK3g95V3w",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "frontsPurgeSqsC6BB4572",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "frontsPurgeSqsC6BB4572",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "mobilefastlycachepurgerFDB90D2A": {
      "DependsOn": [
        "ExecutionRoleDefaultPolicyA5B92313",
        "ExecutionRole605A040B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "mobile-fastly-cache-purger/TEST/mobile-fastly-cache-purger/mobile-fastly-cache-purger.jar",
        },
        "Environment": {
          "Variables": {
            "APP": "mobile-fastly-cache-purger",
            "App": "mobile-fastly-cache-purger",
            "STACK": "mobile-fastly-cache-purger",
            "STAGE": "TEST",
            "Stack": "mobile-fastly-cache-purger",
            "Stage": "TEST",
          },
        },
        "FunctionName": "mobile-fastly-cache-purger-cdk-TEST",
        "Handler": "PurgerLambda::handleRequest",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "ExecutionRole605A040B",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-fastly-cache-purger",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-fastly-cache-purger",
          },
          {
            "Key": "Stack",
            "Value": "mobile-fastly-cache-purger",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "mobilefastlycachepurgerSqsEventSourceMobileFastlyCachePurgerfrontsPurgeSqsE6F58D3DEF06EEF4": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "frontsPurgeSqsC6BB4572",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "mobilefastlycachepurgerFDB90D2A",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
  },
}
`;
